File: wms/apps/android-tsd/app/src/main/java/pro/logoff/wms/tsd/data/OperationEntity.java
package pro.logoff.wms.tsd.data;

import androidx.annotation.NonNull;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

@Entity(
    tableName = "tsd_operations",
    indices = {@Index(value = {"status", "createdAt"})}
)
public class OperationEntity {
    @PrimaryKey
    @NonNull
    public String operationKey;
    public String operationType;
    public String payloadJson;
    public long createdAt;
    public String status;
    public int attempts;
    public String lastMessage;
    public Long lastTriedAt;
    public Long syncedAt;

    public OperationEntity() {
        operationKey = "";
    }

    @Ignore
    public OperationEntity(
        @NonNull String operationKey,
        String operationType,
        String payloadJson,
        long createdAt,
        String status,
        int attempts,
        String lastMessage,
        Long lastTriedAt,
        Long syncedAt
    ) {
        this.operationKey = operationKey;
        this.operationType = operationType;
        this.payloadJson = payloadJson;
        this.createdAt = createdAt;
        this.status = status;
        this.attempts = attempts;
        this.lastMessage = lastMessage;
        this.lastTriedAt = lastTriedAt;
        this.syncedAt = syncedAt;
    }

    public PendingOperation toPendingOperation() {
        return new PendingOperation(
            operationKey,
            operationType,
            PayloadJson.jsonToPayload(payloadJson),
            createdAt,
            OperationStatus.valueOf(status),
            attempts,
            lastMessage
        );
    }

    public static OperationEntity fromPending(PendingOperation operation) {
        return new OperationEntity(
            operation.operationKey,
            operation.operationType,
            PayloadJson.payloadToJson(operation.payload),
            operation.createdAt,
            operation.status.name(),
            operation.attempts,
            operation.lastMessage,
            null,
            null
        );
    }
}

File: wms/apps/android-tsd/app/src/main/java/pro/logoff/wms/tsd/data/OperationEntity.kt
package pro.logoff.wms.tsd.data

import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import org.json.JSONObject

@Entity(
    tableName = "tsd_operations",
    indices = [Index(value = ["status", "createdAt"])],
)
data class OperationEntity(
    @PrimaryKey val operationKey: String,
    val operationType: String,
    val payloadJson: String,
    val createdAt: Long,
    val status: String,
    val attempts: Int,
    val lastMessage: String?,
    val lastTriedAt: Long?,
    val syncedAt: Long?,
) {
    fun toPendingOperation(): PendingOperation =
        PendingOperation(
            operationKey = operationKey,
            operationType = operationType,
            payload = jsonToPayload(payloadJson),
            createdAt = createdAt,
            status = OperationStatus.valueOf(status),
            attempts = attempts,
            lastMessage = lastMessage,
        )

    companion object {
        fun fromPending(operation: PendingOperation): OperationEntity =
            OperationEntity(
                operationKey = operation.operationKey,
                operationType = operation.operationType,
                payloadJson = payloadToJson(operation.payload),
                createdAt = operation.createdAt,
                status = operation.status.name,
                attempts = operation.attempts,
                lastMessage = operation.lastMessage,
                lastTriedAt = null,
                syncedAt = null,
            )
    }
}

// Русский комментарий: payload храним JSON-строкой, чтобы Room не зависел от сетевой модели Retrofit.
fun payloadToJson(payload: Map<String, String>): String = JSONObject(payload).toString()

fun jsonToPayload(payloadJson: String): Map<String, String> {
    val payload = JSONObject(payloadJson)
    return payload.keys().asSequence().associateWith { key -> payload.optString(key) }
}
